lorenz/src/main/java/org/cadixdev/lorenz/merge/MappingSetMergerHandler.java

```diff
@@ -0,0 +1,612 @@
+    }
+
+    /**
+     * Handle the case where only the left inner class mapping exists, creating the result in the {@code target} and
+     * returning it. This method should only map the name of the class, not any members.
+     *
+     * @param left The left class mapping, never {@code null}.
+     * @param target The class mapping to create the new inner class mapping in, never {@code null}.
+     * @param context The {@link MergeContext} associated with this merge operation, never {@code null}.
+     * @return The new inner class mapping in a merge result. Pass {@code null} to the {@link MergeResult} if the
+     *         mapping wasn't merged and should be removed.
+     * @since 0.5.4
+     */
+    default MergeResult<InnerClassMapping> addLeftInnerClassMapping(final InnerClassMapping left, final ClassMapping<?, ?> target, final MergeContext context) {
+        return new MergeResult<>(target.createInnerClassMapping(left.getObfuscatedName(), left.getDeobfuscatedName()));
+    }
+
+    /**
+     * Handle the case where only the right inner class mapping exists, creating the result in the {@code target} and
+     * returning it. This method should only map the name of the class, not any members.
+     *
+     * @param right The right class mapping, never {@code null}.
+     * @param target The class mapping to create the new inner class mapping in, never {@code null}.
+     * @param context The {@link MergeContext} associated with this merge operation, never {@code null}.
+     * @return The new inner class mapping in a merge result. Pass {@code null} to the {@link MergeResult} if the
+     *         mapping wasn't merged and should be removed.
+     * @since 0.5.4
+     */
+    default MergeResult<InnerClassMapping> addRightInnerClassMapping(final InnerClassMapping right, final ClassMapping<?, ?> target, final MergeContext context) {
+        return new MergeResult<>(target.createInnerClassMapping(right.getObfuscatedName(), right.getDeobfuscatedName()), right);
+    }
+
+    /*
+     * Field Mapping Handlers
+     */
+
+    /**
+     * Merge 2 existing fields together, creating the result in the {@code target} and returning it.
+     *
+     * @param left The left field mapping, never {@code null}.
+     * @param right The right field mapping, never {@code null}.
+     * @param target The class mapping to create the new field mapping in, never {@code null}.
+     * @param context The {@link MergeContext} associated with this merge operation, never {@code null}.
+     * @return The new field mapping. Return {@code null} if the mapping wasn't merged and should be removed.
+     * @since 0.5.4
+     */
+    default FieldMapping mergeFieldMappings(
+        final FieldMapping left,
+        final FieldMapping right,
+        final ClassMapping<?, ?> target,
+        final MergeContext context
+    ) {
+        return target.createFieldMapping(left.getObfuscatedName(), right.getDeobfuscatedName());
```

Comment:
This particular method call will create a new `FieldMapping` with a null signature. A relatively big issue with the current merging system is that merging a `FieldSignature` with a type and one without a type won't merge correctly. This appears when merging Yarn / Intermediary mappings with... anything else. Currently, the best solution is to convert Yarn's `FieldSignatures` to have all null types and then merge with your other mapping set, but this obviously eliminates types in your resulting mappings.

I would love to have the `LOOSE` strategy account for this. It would clear up a lot of confusion when merging. Another idea is to have `MethodMergeStrategy` separate from `FieldMergeStrategy` for higher customizability. The `STRICT` strategy should also test for field types to have the same effect as the original API. (The test cases for this could use XSRG.)

Reply:
I wasn't aware there were field mappings with type descriptors, could you give a link to where I could get example mappings to test with so I can play with it? And yeah I agree having merge strategies for fields and methods separate would make sense, having them separate enums (even with identical values) would help with code clarity and documentation.

Reply:
[Here](https://maven.fabricmc.net/net/fabricmc/yarn/1.16.1+build.21/) are the latest Fabric mappings for 1.16.1. `yarn-1.16.1+build.21.jar` is Tiny v1 & `yarn-1.16.1+build.21-v2.jar` is Tiny v2.

[`lorenz-tiny`](https://github.com/FabricMC/lorenz-tiny) can be used to parse it to a `MappingSet`:

```java
TinyMappingFormat.LEGACY.createReader(tinyFile.toPath(), "official", "named").read();
```

Reply:
Field mappings now handle field type signatures with either `LOOSE` (types are ignored) or `STRICT` (types must match) merge strategies - default is `LOOSE`.